hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+    if (path.isEmpty)
+      return path
+
+    if (path.last != '/')
+      return path
+
+    var i = path.length - 1
+    while (i > 0 && path(i - 1) == '/')
+      i -= 1
+    path.substring(0, i)
+  }
+}
+
+class AzureBlobServiceClientCache(credential: TokenCredential) {
+  @transient private lazy val clientBuilder: BlobServiceClientBuilder = new BlobServiceClientBuilder()
+  @transient private lazy val clients: Map[String, BlobServiceClient] = Map()
+
+  def getServiceClient(account: String): BlobServiceClient = {
+    clients.get(account) match {
+      case Some(client) => client
+      case None =>
+        val blobServiceClient = clientBuilder
+          .credential(credential)
+          .endpoint(s"https://$account.blob.core.windows.net")
+          .buildClient()
+        clients + (account -> blobServiceClient)
+        blobServiceClient
+    }
+  }
+}
+
+
+class AzureStorageFS(val serviceAccountKey: Option[String] = None) extends FS {
+  @transient private lazy val storageClientCache = serviceAccountKey match {
+    case None =>
+      val credential: DefaultAzureCredential = new DefaultAzureCredentialBuilder().build()
+      new AzureBlobServiceClientCache(credential)
+    case Some(keyData) =>
+      val kvs = JsonMethods.parse(keyData) match {
+        case json4s.JObject(values) => values.toMap
+      }
+
+      val appId = kvs("appId").asInstanceOf[json4s.JString].s
+      val password = kvs("password").asInstanceOf[json4s.JString].s
+      val tenant = kvs("tenant").asInstanceOf[json4s.JString].s
+
+      val clientSecretCredential: ClientSecretCredential = new ClientSecretCredentialBuilder()
+        .clientId(appId)
+        .clientSecret(password)
+        .tenantId(tenant)
+        .build();
+      new AzureBlobServiceClientCache(clientSecretCredential)
+  }
+
+  def getBlobServiceClient(account: String): BlobServiceClient = {
+    storageClientCache.getServiceClient(account)
+  }
+
+  def getBlobClient(filename: String): BlobClient = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    getBlobServiceClient(account).getBlobContainerClient(container).getBlobClient(path)
+  }
+
+  def getContainerClient(filename: String): BlobContainerClient = {
+    val (account, container, _) = getAccountContainerPath(filename)
+    getBlobServiceClient(account).getBlobContainerClient(container)
+  }
+
+  def openNoCompression(filename: String): SeekableDataInputStream = {
+    val blobClient: BlobClient = getBlobClient(filename)
+
+    val is: SeekableInputStream = new FSSeekableInputStream {
+      private[this] var client: BlobClient = blobClient
+
+      def fill(): Unit = {
+        bb.clear()
+
+        // read some bytes
+        val outputStream: ByteArrayOutputStream = new ByteArrayOutputStream()
+        val blobProperties = client.getProperties
+        val blobSize = blobProperties.getBlobSize
+
+        val count = Math.min(blobSize - pos, bb.remaining() - 1)
```

Comment:
I think I must have confused myself at some point because I thought there would be an off by one error which would cause the buffer to overflow. However I reread the documentation for the [BlobRange class](https://javadoc.io/static/com.azure/azure-storage-blob/12.8.0/com/azure/storage/blob/models/BlobRange.html) and it seems to be what you'd expect where count is the number of bytes being downloaded. I changed it to `bb.remaining()`and all tests are passing so I think that's fine